python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner.py

```diff
@@ -0,0 +1,255 @@
+            self.options.execution_settings
+            or kernel.get_service(self.service_id).get_prompt_execution_settings_class()(service_id=self.service_id)
+        )
+
+        # Clone the kernel so that we can add planner-specific plugins without affecting the original kernel instance
+        cloned_kernel = copy(kernel)
+        cloned_kernel.import_plugin_from_object(UserInteraction(), "UserInteraction")
+
+        # Create and invoke a kernel function to generate the initial plan
+        initial_plan = await self._generate_plan(question, cloned_kernel)
+
+        chat_history_for_steps = await self._build_chat_history_for_step(question, initial_plan, cloned_kernel)
+        prompt_execution_settings.tool_choice = "auto"
+        prompt_execution_settings.tools = get_tool_call_object(kernel, {"exclude_plugin": [self.service_id]})
+        for i in range(self.options.max_iterations):
+            # sleep for a bit to avoid rate limiting
+            if i > 0:
+                await asyncio.sleep(self.options.min_iteration_time_ms / 1000.0)  # convert ms to sec
+            # For each step, request another completion to select a function for that step
+            chat_history_for_steps.add_user_message(STEPWISE_USER_MESSAGE)
+            chat_result = await self._get_completion_with_functions(
+                chat_history_for_steps, cloned_kernel, chat_completion, prompt_execution_settings
+            )
+            chat_result = chat_result[0]
+            chat_history_for_steps.add_message(chat_result)
+
+            if not chat_result.tool_calls:
+                chat_history_for_steps.add_user_message("That function call is invalid. Try something else!")
+                continue
+
+            # Try to get the final answer out
+            if chat_result.tool_calls[0].function.name == USER_INTERACTION_SEND_FINAL_ANSWER:
+                args = json.loads(chat_result.tool_calls[0].function.arguments)
+                answer = args["answer"]
+                return FunctionCallingStepwisePlannerResult(
+                    final_answer=answer,
+                    chat_history=chat_history_for_steps,
+                    iterations=i + 1,
+                )
+
+            chat_history_for_steps = await self._process_tool_calls(chat_result, cloned_kernel, chat_history_for_steps)
+
+        # We're done, but the model hasn't returned a final answer.
+        return FunctionCallingStepwisePlannerResult(
+            final_answer="",
+            chat_history=chat_history_for_steps,
+            iterations=i + 1,
+        )
+
+    async def _process_tool_calls(
+        self,
+        chat_result: OpenAIChatMessageContent,
+        kernel: Kernel,
+        chat_history: ChatHistory,
+    ) -> ChatHistory:
+        """Process the tool calls in the chat result and return the chat history for the steps"""
+        logger.info(f"processing {len(chat_result.tool_calls)} tool calls")
```

Comment:
As is done in dotnet, the tool call handling is done inside the planner as I have here. They're not turning on auto invoke tool calls. We should ask them why.

Reply:
Or are you referring to why not use OpenAIChatCompletionBase's _process_tool_calls method, which we can (so there's no duplicated code?). That can be done. As for auto invoking tools, that seems to be a design choice to manage the tool calls in the planner so that if we get a missing tool, we add a message to try a different function. https://github.com/microsoft/semantic-kernel/blob/555a7c8436214c959550c1ae43e6d774c269df8b/dotnet/src/Planners/Planners.OpenAI/Stepwise/FunctionCallingStepwisePlanner.cs#L129C13-L149C14

Reply:
That makes sense, but catching that should maybe be part of auto invoke anyway? Then we can reuse that, would make this planner simpler and easier to maintain I think? Wdyt?